airbyte-webapp/src/services/connectorBuilder/ConnectorBuilderStateService.tsx

```diff
@@ -148,7 +125,68 @@ export const ConnectorBuilderStateProvider: React.FC<React.PropsWithChildren<unk
   return <ConnectorBuilderStateContext.Provider value={ctx}>{children}</ConnectorBuilderStateContext.Provider>;
 };
 
-export const useConnectorBuilderState = (): Context => {
+export const ConnectorBuilderAPIProvider: React.FC<React.PropsWithChildren<unknown>> = ({ children }) => {
+  const { formatMessage } = useIntl();
+  const { jsonManifest, testStreamIndex, setTestStreamIndex } = useConnectorBuilderState();
+
+  const manifest = jsonManifest ?? DEFAULT_JSON_MANIFEST_VALUES;
+
+  // config
+  const [configString, setConfigString] = useState("{\n  \n}");
+
+  const configJson: StreamReadRequestBodyConfig = useMemo(() => {
+    try {
+      const json = JSON.parse(configString) as StreamReadRequestBodyConfig;
+      return json;
+    } catch (err) {
+      console.error(`Config value is not valid JSON! Error: ${err}`);
+      return {};
+    }
+  }, [configString]);
+
+  // streams
+  const {
+    data: streamListRead,
+    isError: isStreamListError,
+    error: streamListError,
+  } = useListStreams({ manifest, config: configJson });
+  const unknownErrorMessage = formatMessage({ id: "connectorBuilder.unknownError" });
+  const streamListErrorMessage = isStreamListError
+    ? streamListError instanceof Error
+      ? streamListError.message || unknownErrorMessage
+      : unknownErrorMessage
+    : undefined;
+  const streams = useMemo(() => {
+    return streamListRead?.streams ?? [];
+  }, [streamListRead]);
+
+  useEffect(() => {
+    if (testStreamIndex >= streams.length && streams.length > 0) {
+      setTestStreamIndex(streams.length - 1);
+    }
+  }, [streams, testStreamIndex, setTestStreamIndex]);
+
+  const ctx = {
+    streams,
+    streamListErrorMessage,
+    configJson,
+    configString,
+    setConfigString,
```

Comment:
nit: I feel like we should rename this to something like `testInputJson`, since I think `config` may start to get too overloaded. Not super related to the other changes here though